server/src/main/java/org/opensearch/index/query/SourceFieldMatchQuery.java

```diff
@@ -73,7 +73,7 @@ public void visit(QueryVisitor visitor) {
 
     @Override
     public Query rewrite(IndexSearcher indexSearcher) throws IOException {
-        Query rewritten = indexSearcher.rewrite(delegateQuery);
+        Query rewritten = delegateQuery.rewrite(indexSearcher);
```

Comment:
@rishabhmaurya I was reading the description and it sounds like the current behavior (indexSearcher::rewrite) is correct: the `profile` (+ timers) is also measuring the rewrite times (if I am not mistaken) and it seems like we want to phase this part out, does not look right.

Reply:
you're right however in query phase, when rewrite is called

https://github.com/opensearch-project/OpenSearch/blob/5ff50a7e00698b0f448bc4840461c8518500d591/server/src/main/java/org/opensearch/search/query/QueryPhase.java#L127

the `ContextIndexSearch` will start the timer for the query and will start noting the `rewrite()` time for all clauses in the clauses for the complete query tree. 

But if we try to use `indexSearcher.rewrite(delegateQuery)` within rewrite of any of the query, it will try to initialize the profiler and timer again and will result in the assertion error here - 

https://github.com/opensearch-project/OpenSearch/blob/200ad5d28a577877be530ecab507601898025c5c/server/src/main/java/org/opensearch/search/profile/query/AbstractQueryProfileTree.java#L48

```
»  java.lang.AssertionError
»       at org.opensearch.search.profile.query.AbstractQueryProfileTree.startRewriteTime(AbstractQueryProfileTree.java:48)
»       at org.opensearch.search.profile.query.QueryProfiler.startRewriteTime(QueryProfiler.java:80)
»       at org.opensearch.search.internal.ContextIndexSearcher.rewrite(ContextIndexSearcher.java:192)
»       at org.opensearch.index.query.SourceFieldMatchQuery.rewrite(SourceFieldMatchQuery.java:76)
»       at org.apache.lucene.search.IndexSearcher.rewrite(IndexSearcher.java:777)
»       at org.opensearch.search.internal.ContextIndexSearcher.rewrite(ContextIndexSearcher.java:196)
»       at org.opensearch.search.DefaultSearchContext.preProcess(DefaultSearchContext.java:362)
»       at org.opensearch.search.query.QueryPhase.preProcess(QueryPhase.java:127)
»       at org.opensearch.search.SearchService.createContext(SearchService.java:1053)
```

Reply:
let me also update the description of the PR

Reply:
> @reta you're right however in query phase, when rewrite is called

Oh I see, thank you @rishabhmaurya ! Could we come up with the test for this flow to capture any such flaws early? We do have many tests for profile, did we miss this particular query type?

Reply:
@reta sure! could you point me to a unit test covering profile for any of the field type? 
Otherwise I can write a rest api spec test to cover it.

Reply:
> @reta sure! could you point me to a unit test covering profile for any of the field type?

Sure @rishabhmaurya , please look for `Q*Profile*Test`, `*Profile*Test`, there are quite a few

Reply:
thanks @reta, writing a test now. Unfortunately, I don't think I will be able to meet the 4pm PT deadline for 2.15 code freeze due to my busy oncall today :(

Reply:
I think we still could ask for merge if needed: I am concerned that a) currently all tests pass and b) with this change all tests pass, it is unclear what is happening without tests.

Reply:
@reta I have added a test to cover this scenario. Please take a look

Also, for most of the field types and their query, we don't have unit tests with profile enabled and this was one of them. But I agree that going forward we should add such tests whenever introducing a new query or a field type.